---
output:
  md_document:
    variant: markdown_github
---
<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# grur <img src="docs/logo.png" align="right" alt="" />
<!-- badges: start -->
```{r, echo = FALSE}
description <- readLines("DESCRIPTION")
rvers <- stringr::str_match(grep("R \\(", description, value = TRUE), "[0-9]{1,4}\\.[0-9]{1,4}\\.[0-9]{1,4}")[1,1]
version <- gsub(" ", "", gsub("Version:", "", grep("Version:", description, value = TRUE)))
```
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://tidyverse.org/lifecycle/#experimental)
[![Travis-CI Build Status](https://travis-ci.org/thierrygosselin/grur.svg?branch=master)](https://travis-ci.org/thierrygosselin/grur)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/thierrygosselin/grur?branch=master&svg=true)](https://ci.appveyor.com/project/thierrygosselin/grur)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/grur)](http://cran.r-project.org/package=grur)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![DOI](https://zenodo.org/badge/87596763.svg)](https://zenodo.org/badge/latestdoi/87596763)
[![packageversion](https://img.shields.io/badge/Package%20version-`r version`-orange.svg)](commits/master)
[![Last-changedate](https://img.shields.io/badge/last%20change-`r gsub('-', '--', Sys.Date())`-brightgreen.svg)](/commits/master)
<!-- badges: end -->
---


https://thierrygosselin.github.io/grur/

The name **grur** |ɡro͞oˈr| was chosen because the missing genotypes dilemma with
RADseq data reminds me of the cheese paradox. 

Here, I don't want to sustain a
[war](http://www.lefigaro.fr/flash-eco/2012/12/07/97002-20121207FILWWW00487-le-gruyere-francais-doit-avoir-des-trous.php)
or the controversy of cheese with holes,
so choose as you like, the *French Gruyère* or the *Swiss Emmental*.
The paradox is that the more cheese you have the more holes you'll get.
But, the more holes you have means the less cheese you have...
So, someone could conclude, the more cheese = the less cheese ?
I'll leave that up to you, back to genomics...

Numerous genomic analysis are vulnerable to missing values,
don't get trapped by missing genotypes in your RADseq dataset.

Use **grur** to **visualize patterns of missingness** and 
**perform map-independent imputations of missing genotypes** 
(see [features](https://github.com/thierrygosselin/grur#features) below).



## Installation
```r
if (!require("remotes")) install.packages("remotes")
remotes::install_github("thierrygosselin/grur")
library(grur)
```
Note: not all the packages used inside **grur** are installed, why?

* Depending on your analysis, you might not need to install all of them
* Some modules used for imputations are more complicated to install, and might
impact your choice of imputations.
* [this section covers all the required packages installation](https://thierrygosselin.github.io/grur/index.html#options-and-required-packages)

## web site and additional info:

* [https://thierrygosselin.github.io/grur/](https://thierrygosselin.github.io/grur/) 
* [Computer setup and troubleshooting](http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html)
* [Function's documentation](http://thierrygosselin.github.io/grur/reference/index.html)
* [Vignettes](http://thierrygosselin.github.io/grur/articles/index.html)
* How to cite grur: inside R type `citation("grur")`

## [Life cycle](https://thierrygosselin.github.io/radiator/articles/life_cycle.html)

**grur** is still experimental, in order to make the package better, changes are 
inevitable. Experimental functions will change, argument names will change.
Your codes and workflows might break from time to time **until grur is stable**. 
Consequently, depending on your tolerance to change, **grur** might not be for you.

* Philosophy, major changes and deprecated functions/arguments are documented in
life cycle section of functions. 
* [changelog, versions, new features and bug history](https://thierrygosselin.github.io/grur/news/index.html)
* [issues](https://github.com/thierrygosselin/grur/issues/new/choose) and [contributions](https://github.com/thierrygosselin/grur/issues/new/choose)


## Options and required packages

Please follow additional instructions in the [vignette](http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html) to install the required packages for the imputation options you want to conduct:

| imputation options | package | install instructions |
|:-------------------|:-------------------------:|-------------------------------|
|**imputation.method = "lightgbm"**|`lightgbm`|[vignette](http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html)|
|**imputation.method = "xgboost"**|`xgboost`|[vignette](http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html)|
|**imputation.method = "rf"**|`randomForestSRC`|[vignette](http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html)|
|**imputation.method = "rf_pred"**|`ranger`|`install.packages("ranger")`|
|**if using pmm > 0**|`missRanger`|`install.packages("missRanger")`|


## Features
| Caracteristics | Description |
|:-------------------|:--------------------------------------------------------|
|**Simulate RADseq data**|`simulate_rad`: simulate populations of RADseq data following island or stepping stone models. Inside the function, allele frequency can be created with [fastsimcoal2](http://cmpg.unibe.ch/software/fastsimcoal2/) and then used inside [rmetasim](https://github.com/stranda/rmetasim) simulation engine. *Vignette coming soon*.|
|**Patterns of missingness**|`missing_visualization`: visualize patterns of missing data associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, homozygosity, etc). Similar to PLINK's identify-by-missingness analysis (IBM), **grur** is more powerful because it generates more analysis and automatically creates tables and figures ([see vignette](https://thierrygosselin.github.io/grur/articles/vignette_missing_data_analysis.html)). <br><br>`generate_missing`: allows to generate missing genotypes in dateset [simulated] based on a compound Dirichlet-multinomial distribution. *Vignette coming soon*.|
|**Imputations**|`grur_imputations`: **Map-independent** imputations of missing genotypes with several algorithms (including machine leaning):<br> * **Random Forests** (on-the-fly-imputations with randomForestSRC or using predictive modelling using ranger and missRanger),<br>* **Extreme Gradient Tree Boosting** (using XGBoost or LightGBM),<br>* **Bayesian PCA** (using bpca in pcaMethods),<br>* **Classic Strawman: ** the most frequently observed, non-missing, genotypes is used for imputation.<br><br>**Hierarchy: ** algorithm's model can account for *strata* groupings, e.g. if patterns of missingness is found in the data.<br><br>**Haplotypes: ** automatically detect SNPs on the same LOCUS (read/haplotype) to impute the SNPs jointly, reducing imputation artifacts. *Vignette coming soon*.|
| **Input/Output** | **grur** uses [radiator](https://thierrygosselin.github.io/radiator/index.html) input and output modules. Check out the [overview](https://thierrygosselin.github.io/radiator/articles/get_started.html#overview) of
supported file format|
|**[ggplot2](http://ggplot2.org)-based plotting**|Visualization: publication-ready figures of important metrics and statistics.|
|**Parallel**|Codes designed and optimized for fast computations with, sometimes, progress bars. Works with all OS: Linux, Mac and yes PC!|
