<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Imputations of RADseq data • grur</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Imputations of RADseq data">
<meta property="og:description" content="RADseq data imputations.">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">grur</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/rad_genomics_computer_setup.html">RADseq genomics in R</a>
    </li>
    <li>
      <a href="articles/vignette_missing_data_analysis.html">grur missing data visualization analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/grur">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="grur" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#grur" class="anchor"></a>grur<img src="../../../../Mobile%20Documents/com~apple~CloudDocs/r_packages/grur/logo.png" align="right">
</h1></div>
<!-- badges: start -->
<p><a href="https://tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="lifecycle"></a> <a href="https://travis-ci.org/thierrygosselin/grur"><img src="https://travis-ci.org/thierrygosselin/grur.svg?branch=master" alt="Travis-CI Build Status"></a> <a href="https://ci.appveyor.com/project/thierrygosselin/grur"><img src="https://ci.appveyor.com/api/projects/status/github/thierrygosselin/grur?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a> <a href="http://cran.r-project.org/package=grur"><img src="http://www.r-pkg.org/badges/version/grur" alt="CRAN_Status_Badge"></a> <a href="http://www.repostatus.org/#active"><img src="http://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a> <a href="https://zenodo.org/badge/latestdoi/87596763"><img src="https://zenodo.org/badge/87596763.svg" alt="DOI"></a> <a href="commits/master"><img src="https://img.shields.io/badge/Package%20version-0.1.0-orange.svg" alt="packageversion"></a> <a href="/commits/master"><img src="https://img.shields.io/badge/last%20change-2019--04--30-brightgreen.svg" alt="Last-changedate"></a> <!-- badges: end --> —</p>
<p><a href="https://thierrygosselin.github.io/grur/" class="uri">https://thierrygosselin.github.io/grur/</a></p>
<p>The name <strong>grur</strong> |ɡro͞oˈr| was chosen because the missing genotypes dilemma with RADseq data reminds me of the cheese paradox.</p>
<p>Here, I don’t want to sustain a <a href="http://www.lefigaro.fr/flash-eco/2012/12/07/97002-20121207FILWWW00487-le-gruyere-francais-doit-avoir-des-trous.php">war</a> or the controversy of cheese with holes, so choose as you like, the <em>French Gruyère</em> or the <em>Swiss Emmental</em>. The paradox is that the more cheese you have the more holes you’ll get. But, the more holes you have means the less cheese you have… So, someone could conclude, the more cheese = the less cheese ? I’ll leave that up to you, back to genomics…</p>
<p>Numerous genomic analysis are vulnerable to missing values, don’t get trapped by missing genotypes in your RADseq dataset.</p>
<p>Use <strong>grur</strong> to <strong>visualize patterns of missingness</strong> and <strong>perform map-independent imputations of missing genotypes</strong> (see <a href="https://github.com/thierrygosselin/grur#features">features</a> below).</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To try out the dev version of <strong>grur</strong>, copy/paste the code below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(<span class="st">"remotes"</span>)) <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"remotes"</span>)</a>
<a class="sourceLine" id="cb1-2" title="2">remotes<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/remotes/topics/install_github">install_github</a></span>(<span class="st">"thierrygosselin/grur"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(grur)</a></code></pre></div>
</div>
<div id="options-and-required-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#options-and-required-packages" class="anchor"></a>Options and required packages</h2>
<p>Please follow instructions in the <a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">Notebook vignette</a> to install required packages for the selected imputation options below:</p>
<table class="table table">
<colgroup>
<col width="25%">
<col width="34%">
<col width="39%">
</colgroup>
<thead><tr class="header">
<th align="left">imputation options</th>
<th align="center">package</th>
<th>install instructions</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>imputation.method = “lightgbm”</strong></td>
<td align="center"><code>lightgbm</code></td>
<td><a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">Notebook vignette</a></td>
</tr>
<tr class="even">
<td align="left"><strong>imputation.method = “xgboost”</strong></td>
<td align="center"><code>xgboost</code></td>
<td><a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">Notebook vignette</a></td>
</tr>
<tr class="odd">
<td align="left"><strong>imputation.method = “rf”</strong></td>
<td align="center"><code>randomForestSRC</code></td>
<td><a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">Notebook vignette</a></td>
</tr>
<tr class="even">
<td align="left"><strong>imputation.method = “rf_pred”</strong></td>
<td align="center"><code>ranger</code></td>
<td><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("ranger")</a></code></td>
</tr>
<tr class="odd">
<td align="left"><strong>imputation.method = “bpca”</strong></td>
<td align="center"><code>pcaMethods</code></td>
<td>
<code><a href="https://www.rdocumentation.org/packages/base/topics/source">source("https://bioconductor.org/biocLite.R")</a></code><br><code>biocLite("pcaMethods")</code>
</td>
</tr>
<tr class="even">
<td align="left"><strong>if using pmm &gt; 0</strong></td>
<td align="center"><code>missRanger</code></td>
<td><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("missRanger")</a></code></td>
</tr>
</tbody>
</table>
<div id="troubleshooting" class="section level4">
<h4 class="hasAnchor">
<a href="#troubleshooting" class="anchor"></a>Troubleshooting</h4>
<ul>
<li>
<code>rmetasim</code> needs to be modified in order to simulate more than 2000 markers <a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">notebook vignette</a>
</li>
<li>
<strong>Parallel computing</strong>: follow the steps in this <a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">notebook vignette</a> to install the packages with OpenMP-enabled compiler and conduct imputations in parallel.</li>
<li><a href="https://www.dropbox.com/s/5npumwdo0cxtxi4/rad_genomics_computer_setup.nb.html?dl=0">Installation problems.</a></li>
<li>
<strong>Windows users</strong>: Install <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a>.</li>
<li>The R GUI is unstable with functions using parallel (<a href="https://stat.ethz.ch/R-manual/R-devel/library/parallel/html/mclapply.html">more info</a>), so I recommend using <a href="https://www.rstudio.com/products/rstudio/download/">RStudio</a> for a better experience.</li>
<li>Running codes in chunks inside R Notebook might cause problem, run it outside in the console.</li>
</ul>
</div>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<table class="table table">
<colgroup>
<col width="25%">
<col width="74%">
</colgroup>
<thead><tr class="header">
<th align="left">Caracteristics</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Simulate RADseq data</strong></td>
<td align="left">
<code>simulate_rad</code>: simulate populations of RADseq data following island or stepping stone models. Inside the function, allele frequency can be created with <a href="http://cmpg.unibe.ch/software/fastsimcoal2/">fastsimcoal2</a> and then used inside <a href="https://github.com/stranda/rmetasim">rmetasim</a> simulation engine. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="even">
<td align="left"><strong>Patterns of missingness</strong></td>
<td align="left">
<code>missing_visualization</code>: visualize patterns of missing data associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, homozygosity, etc). Similar to PLINK’s identify-by-missingness analysis (IBM), <strong>grur</strong> is more powerful because it generates more analysis and automatically creates tables and figures. Vignette: <a href="https://www.dropbox.com/s/4zf032g6yjatj0a/vignette_missing_data_analysis.nb.html?dl=0">html</a> and <a href="https://www.dropbox.com/s/5fxw2h9w1l1j391/vignette_missing_data_analysis.Rmd?dl=0">Rmd</a><br><br><code>generate_missing</code>: allows to generate missing genotypes in dateset [simulated] based on a compound Dirichlet-multinomial distribution. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="odd">
<td align="left"><strong>Imputations</strong></td>
<td align="left">
<code>grur_imputations</code>: <strong>Map-independent</strong> imputations of missing genotypes with several algorithms (including machine leaning):<br> * <strong>Random Forests</strong> (on-the-fly-imputations with randomForestSRC or using predictive modelling using ranger and missRanger),<br>* <strong>Extreme Gradient Tree Boosting</strong> (using XGBoost or LightGBM),<br>* <strong>Bayesian PCA</strong> (using bpca in pcaMethods),<br>* <strong>Classic Strawman: </strong> the most frequently observed, non-missing, genotypes is used for imputation.<br><br><strong>Hierarchy: </strong> algorithm’s model can account for <em>strata</em> groupings, e.g. if patterns of missingness is found in the data.<br><br><strong>Haplotypes: </strong> automatically detect SNPs on the same LOCUS (read/haplotype) to impute the SNPs jointly, reducing imputation artifacts. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="even">
<td align="left"><strong>Input/Output</strong></td>
<td align="left">The imputations offered in <strong>grur</strong> are seamlesly integrated in <strong>radiator</strong> and <strong>assigner</strong>. Imputations are also integrated with usefull filters, blacklists and whitelists inside those 2 packages. Genetic formats recognized: <a href="https://samtools.github.io/hts-specs/">VCF, SNPs and haplotypes</a>, <a href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml#tr">PLINK tped/tfam</a>, <a href="https://github.com/thibautjombart/adegenet">genind</a>, <a href="https://github.com/thibautjombart/adegenet">genlight</a>, <a href="https://github.com/EricArcher/strataG">strataG gtypes</a>, <a href="http://genepop.curtin.edu.au">Genepop</a>, <a href="http://catchenlab.life.illinois.edu/stacks/">STACKS haplotype file</a>, <a href="https://github.com/jgx65/hierfstat">hierfstat</a>, <a href="https://www.zsl.org/science/software/colony">COLONY</a>, <a href="http://adn.biol.umontreal.ca/~numericalecology/Rcode/">betadiv</a>, <a href="http://gutengroup.mcb.arizona.edu/software/">δaδi</a>, <a href="http://pritchardlab.stanford.edu/structure.html">structure</a>, <a href="http://cmpg.unibe.ch/software/arlequin35/">Arlequin</a>, <a href="https://github.com/zhengxwen/SNPRelate">SNPRelate</a>, dataframes of genotypes in wide or long/tidy format.</td>
</tr>
<tr class="odd">
<td align="left"><strong><a href="http://ggplot2.org">ggplot2</a>-based plotting</strong></td>
<td align="left">Visualization: publication-ready figures of important metrics and statistics.</td>
</tr>
<tr class="even">
<td align="left"><strong>Parallel</strong></td>
<td align="left">Codes designed and optimized for fast computations with progress bars. Works with all OS: Linux, Mac and yes PC!</td>
</tr>
</tbody>
</table>
</div>
<div id="vignettes-and-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes-and-examples" class="anchor"></a>Vignettes and examples</h2>
<p>Vignettes with real data for example in the form of R Notebooks take too much space to be included in package, without CRAN complaining. Consequently, vignettes will start to be distributed separately, follow the links below.</p>
<ul>
<li>Missing data visualization and analysis <a href="https://www.dropbox.com/s/4zf032g6yjatj0a/vignette_missing_data_analysis.nb.html?dl=0">(html vignette)</a> and <a href="https://www.dropbox.com/s/5fxw2h9w1l1j391/vignette_missing_data_analysis.Rmd?dl=0">(Rmd)</a>
</li>
</ul>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>To get the citation, inside R:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/citation">citation</a></span>(<span class="st">"grur"</span>)</a></code></pre></div>
</div>
<div id="new-features" class="section level2">
<h2 class="hasAnchor">
<a href="#new-features" class="anchor"></a>New features</h2>
<p>Change log, version, new features and bug history lives in the <a href="https://github.com/thierrygosselin/grur/blob/master/NEWS.md">NEWS.md file</a></p>
<p><strong>grur v.0.0.10 2018-04-26</strong></p>
<p><code>grur's</code> dependencies:</p>
<ul>
<li>I transferred to <code>Suggests</code> section these packages: lightgbm, missRanger, randomForestSRC, ranger, rmarkdown, rmetasim, strataG, xgboost.</li>
<li>Functions thate requires specific package will now say so.</li>
<li>Reason: people only interested in <code>missing_visualization</code> don’t have to install all the required packages required for imputations or simulations.</li>
</ul>
<p><code>simulate_rad</code>: with the latest R release (3.5.0), Check now throw a new note: <strong>Note: next used in wrong context: no loop is visible at simulate_rad.R:189</strong> I replaced <code>next</code> inside <code>sapply</code> with <code>while</code>.</p>
<p><strong>grur v.0.0.9 2017-10-27</strong></p>
<ul>
<li>
<code>lightGBM</code> option to conduct the imputations is fully functional</li>
</ul>
<div id="roadmap-of-future-developments" class="section level4">
<h4 class="hasAnchor">
<a href="#roadmap-of-future-developments" class="anchor"></a>Roadmap of future developments</h4>
<ul>
<li>Integrate more imputation method.</li>
<li>Workflow tutorial to further explore some problems.</li>
<li>Use Shiny and ggvis (when subplots and/or facets becomes available for ggvis).</li>
<li>Until publication <strong>grur</strong> will change rapidly, stay updated with releases and contribute with bug reports.</li>
<li>Suggestions ?</li>
</ul>
</div>
<div id="contributions" class="section level4">
<h4 class="hasAnchor">
<a href="#contributions" class="anchor"></a>Contributions</h4>
<p>This package has been developed in the open, and it wouldn’t be nearly as good without your contributions. There are a number of ways you can help me make this package even better:</p>
<ul>
<li>If you don’t understand something, please let me know and raise an <a href="https://github.com/thierrygosselin/grur/issues">issue</a>
</li>
<li>Your feedback on what is confusing or hard to understand is valuable.</li>
<li>If you spot a typo, feel free to edit the underlying page and send a pull request.</li>
<li>New to pull request on github ? <a href="http://r-pkgs.had.co.nz/git.html#git-pullreq">The process is very easy</a>.</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/thierrygosselin/grur">https://​github.com/​thierrygosselin/​grur</a>
</li>
<li>Report a bug at <br><a href="https://github.com/thierrygosselin/grur/issues">https://​github.com/​thierrygosselin/​grur/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=2)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing grur</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Thierry Gosselin <br><small class="roles"> Author, maintainer </small>  </li>
<li>Eric Archer <br><small class="roles"> Author </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin, Eric Archer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
