<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Imputations of RADseq data • grur</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Imputations of RADseq data">
<meta property="og:description" content="RADseq data imputations.">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">grur</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">Installation</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#assumptions-before-imputing-your-dataset">Assumptions</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#features">Features</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="articles/vignette_missing_data_analysis.html">Missing visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/grur">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="grur" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#grur" class="anchor"></a>grur <img src="docs/logo.png" align="right" alt="">
</h1></div>
<!-- badges: start -->

<p><a href="https://thierrygosselin.github.io/grur/" class="uri">https://thierrygosselin.github.io/grur/</a></p>
<p>The name <strong>grur</strong> |ɡro͞oˈr| was chosen because the missing genotypes dilemma with RADseq data reminds me of the cheese paradox.</p>
<p>Here, I don’t want to sustain a <a href="http://www.lefigaro.fr/flash-eco/2012/12/07/97002-20121207FILWWW00487-le-gruyere-francais-doit-avoir-des-trous.php">war</a> or the controversy of cheese with holes, so choose as you like, the <em>French Gruyère</em> or the <em>Swiss Emmental</em>. The paradox is that the more cheese you have the more holes you’ll get. But, the more holes you have means the less cheese you have… So, someone could conclude, the more cheese = the less cheese ? I’ll leave that up to you, back to genomics…</p>
<p>Numerous genomic analysis are vulnerable to missing values, don’t get trapped by missing genotypes in your RADseq dataset.</p>
<p>Use <strong>grur</strong> to <strong>visualize patterns of missingness</strong> and <strong>perform map-independent imputations of missing genotypes</strong> (see <a href="https://github.com/thierrygosselin/grur#features">features</a> below).</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"remotes"</span>)) <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"remotes"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">remotes<span class="op">::</span><span class="kw"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">"thierrygosselin/grur"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(grur)</a></code></pre></div>
<p>Note: not all the packages used for imputations inside <strong>grur</strong> are installed automatically, why?</p>
<ul>
<li>Not all methods will be of interest.</li>
<li>Some modules used for imputations are more complicated to install, and depending on OS, it will definitely test your R skills and patience.</li>
<li>By default, you’ll be able to run <code><a href="reference/missing_visualization.html">grur::missing_visualization</a></code> to check for pattern of missingness (the first step…)</li>
</ul>
</div>
<div id="installation-details-for-additonal-imputation-options" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-details-for-additonal-imputation-options" class="anchor"></a>Installation details for additonal imputation options</h2>
<p>Please follow additional instructions in the <a href="http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html">vignette</a> to install the required packages for the imputation options you want to conduct:</p>
<table class="table">
<thead><tr class="header">
<th align="left">imputation options</th>
<th align="center">package</th>
<th align="center">installation difficulty</th>
<th align="left">install instructions</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>imputation.method = “lightgbm”</strong></td>
<td align="center"><code>lightgbm</code></td>
<td align="center">difficult</td>
<td align="left"><a href="http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html">vignette</a></td>
</tr>
<tr class="even">
<td align="left"><strong>imputation.method = “xgboost”</strong></td>
<td align="center"><code>xgboost</code></td>
<td align="center">moderate</td>
<td align="left"><a href="http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html">vignette</a></td>
</tr>
<tr class="odd">
<td align="left"><strong>imputation.method = “rf”</strong></td>
<td align="center"><code>randomForestSRC</code></td>
<td align="center">moderate</td>
<td align="left"><a href="http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html">vignette</a></td>
</tr>
<tr class="even">
<td align="left"><strong>imputation.method = “rf_pred”</strong></td>
<td align="center"><code>ranger</code></td>
<td align="center">easy</td>
<td align="left"><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("ranger")</a></code></td>
</tr>
<tr class="odd">
<td align="left"><strong>if using pmm &gt; 0</strong></td>
<td align="center"><code>missRanger</code></td>
<td align="center">easy</td>
<td align="left"><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("missRanger")</a></code></td>
</tr>
</tbody>
</table>
<p>web site and additional info: <a href="https://thierrygosselin.github.io/grur/" class="uri">https://thierrygosselin.github.io/grur/</a></p>
<ul>
<li><a href="http://thierrygosselin.github.io/grur/articles/rad_genomics_computer_setup.html">Computer setup - installation - troubleshooting</a></li>
<li><a href="http://thierrygosselin.github.io/grur/reference/index.html">Function’s documentation</a></li>
<li><a href="https://thierrygosselin.github.io/grur/index.html#features">grur’s features</a></li>
<li><a href="http://thierrygosselin.github.io/grur/articles/index.html">Vignettes</a></li>
<li>How to cite grur: inside R type <code><a href="https://rdrr.io/r/utils/citation.html">citation("grur")</a></code>
</li>
</ul>
</div>
<div id="life-cycle" class="section level2">
<h2 class="hasAnchor">
<a href="#life-cycle" class="anchor"></a>Life cycle</h2>
<p><strong>grur</strong> is still experimental, in order to make the package better, changes are inevitable. Experimental functions will change, argument names will change. Your codes and workflows might break from time to time <strong>until grur is stable</strong>. Consequently, depending on your tolerance to change, <strong>grur</strong> might not be for you.</p>
<ul>
<li>Philosophy, major changes and deprecated functions/arguments are documented in life cycle section of functions.</li>
<li><a href="https://thierrygosselin.github.io/grur/news/index.html">changelog, versions, new features and bug history</a></li>
<li>
<a href="https://github.com/thierrygosselin/grur/issues/new/choose">issues</a> and <a href="https://github.com/thierrygosselin/grur/issues/new/choose">contributions</a>
</li>
</ul>
</div>
<div id="assumptions-before-imputing-your-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#assumptions-before-imputing-your-dataset" class="anchor"></a>Assumptions before imputing your dataset</h2>
<ol>
<li><p><strong>Filtered data</strong>: Please don’t try <strong>grur</strong> with raw data consisting of &gt; 100K SNPs, you will generate all sorts of bias and you’ll be disapointed. <strong>Filter your data first!</strong> <a href="https://thierrygosselin.github.io/radiator/">radiator</a> was designed for this.</p></li>
<li><p><strong>Correlations</strong>: Machine learning algorithms will work better and faster if correlations are reduced to a minimum. If you used <a href="https://thierrygosselin.github.io/radiator/reference/filter_rad.html">filter_rad</a> to filter your dataset, you should be ok. If not, check your dataset for <a href="https://thierrygosselin.github.io/radiator/reference/filter_ld.html">short and long LD</a>.</p></li>
<li><p><strong>Pattern of individual heterozygosity</strong>: If you have individual heterozygosity patterns and/or correlation of individual heterozygosity with missingness, you might want to skip imputation and go back to filter your data. Please check out <a href="https://thierrygosselin.github.io/radiator/reference/detect_mixed_genomes.html">radiator::detect_mixed_genomes</a> and <a href="https://thierrygosselin.github.io/radiator/reference/detect_het_outliers.html">radiator::detect_het_outliers</a>.</p></li>
<li><p><strong>Patterns of missingness</strong>: Look for patterns of missingness <a href="https://thierrygosselin.github.io/grur/articles/vignette_missing_data_analysis.html">(vignette)</a> to better understand the reasons for their presence and tailor the arguments inside grur’s imputation module.</p></li>
<li><p><strong>Default arguments</strong>: Defaults are there for testing, please, please, please, don’t use grur’s defaults for publications!</p></li>
</ol>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<table class="table">
<colgroup>
<col width="7%">
<col width="92%">
</colgroup>
<thead><tr class="header">
<th align="left">Caracteristics</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>Simulate RADseq data</strong></td>
<td align="left">
<code>simulate_rad</code>: simulate populations of RADseq data following island or stepping stone models. Inside the function, allele frequency can be created with <a href="http://cmpg.unibe.ch/software/fastsimcoal2/">fastsimcoal2</a> and then used inside <a href="https://github.com/stranda/rmetasim">rmetasim</a> simulation engine. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="even">
<td align="left"><strong>Patterns of missingness</strong></td>
<td align="left">
<code>missing_visualization</code>: visualize patterns of missing data associated with different variables of your study (lanes, chips, sequencers, populations, sample sites, reads/samples, homozygosity, etc). Similar to PLINK’s identify-by-missingness analysis (IBM), <strong>grur</strong> is more powerful because it generates more analysis and automatically creates tables and figures (<a href="https://thierrygosselin.github.io/grur/articles/vignette_missing_data_analysis.html">see vignette</a>). <br><br><code>generate_missing</code>: allows to generate missing genotypes in dataset [simulated] based on a compound Dirichlet-multinomial distribution. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="odd">
<td align="left">
<strong>Imputations</strong> <code>grur_imputations</code>
</td>
<td align="left">
<strong>Map-independent</strong> imputations of missing genotypes with several algorithms (including machine leaning):<br>* <strong>Random Forests</strong> (on-the-fly-imputations with randomForestSRC or using predictive modelling using ranger and missRanger),<br>* <strong>Extreme Gradient Tree Boosting</strong> (using XGBoost or LightGBM),<br>* <strong>Bayesian PCA</strong> (using bpca in pcaMethods),<br>* <strong>Classic Strawman:</strong> the most frequently observed, non-missing, genotypes is used for imputation.<br><br><strong>Hierarchy:</strong> algorithm’s model can account for <em>strata</em> groupings, e.g. if patterns of missingness is found in the data.<br><br><strong>Haplotypes:</strong> automatically detect SNPs on the same LOCUS (read/haplotype) to impute the SNPs jointly, reducing imputation artifacts. <em>Vignette coming soon</em>.</td>
</tr>
<tr class="even">
<td align="left"><strong>Input/Output</strong></td>
<td align="left">
<strong>grur</strong> uses <a href="https://thierrygosselin.github.io/radiator/index.html">radiator</a> input and output modules. Check out the <a href="https://thierrygosselin.github.io/radiator/articles/get_started.html#overview">overview</a> of supported file format.</td>
</tr>
<tr class="odd">
<td align="left"><strong><a href="https://ggplot2.tidyverse.org">ggplot2</a>-based plotting</strong></td>
<td align="left">Visualization: publication-ready figures of important metrics and statistics.</td>
</tr>
<tr class="even">
<td align="left"><strong>Parallel</strong></td>
<td align="left">Codes designed and optimized for fast computations with, sometimes, progress bars. Works with all OS: Linux, Mac and yes PC!</td>
</tr>
</tbody>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/thierrygosselin/grur">https://​github.com/​thierrygosselin/​grur</a>
</li>
<li>Report a bug at <br><a href="https://github.com/thierrygosselin/grur/issues">https://​github.com/​thierrygosselin/​grur/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=2)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing grur</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Thierry Gosselin <br><small class="roles"> Author, maintainer </small>  </li>
<li>Eric Archer <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="lifecycle"></a></li>
<li><a href="https://travis-ci.org/thierrygosselin/grur"><img src="https://travis-ci.org/thierrygosselin/grur.svg?branch=master" alt="Travis-CI Build Status"></a></li>
<li><a href="http://cran.r-project.org/package=grur"><img src="http://www.r-pkg.org/badges/version/grur" alt="CRAN_Status_Badge"></a></li>
<li><a href="http://www.repostatus.org/#active"><img src="http://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/87596763"><img src="https://zenodo.org/badge/87596763.svg" alt="DOI"></a></li>
<li><a href="commits/master"><img src="https://img.shields.io/badge/Package%20version-0.1.2-orange.svg" alt="packageversion"></a></li>
<li><a href="/commits/master"><img src="https://img.shields.io/badge/last%20change-2020--03--04-brightgreen.svg" alt="Last-changedate"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin, Eric Archer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
