<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>grur missing data visualization analysis • grur</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="grur missing data visualization analysis">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">grur</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#assumptions-before-imputing-your-dataset">Assumptions</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#features">Features</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/vignette_missing_data_analysis.html">Missing visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/thierrygosselin/grur">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>grur missing data visualization analysis</h1>
                        <h4 class="author">Thierry Gosselin</h4>
            
            <h4 class="date">2019-04-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/thierrygosselin/grur/blob/master/vignettes/vignette_missing_data_analysis.Rmd"><code>vignettes/vignette_missing_data_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>vignette_missing_data_analysis.Rmd</code></div>

    </div>

    
    
<div id="objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#objectives" class="anchor"></a>Objectives</h1>
<p>Learn how to visualize missing genotypes in your genomic dataset with the function <code><a href="../reference/missing_visualization.html">grur::missing_visualization</a></code> (time = 15 min).</p>
</div>
<div id="workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h1>
<p>The function <code>missing_visualization</code> in <a href="https://github.com/thierrygosselin/grur">grur</a> uses various genomic input files and conduct identity-by-missingness analyses (IBM) using Principal Coordinates Analysis (PCoA), also called Multidimensional Scaling (MDS) and RDA (Redundancy Analysis) to highlight missing data patterns. Figures and summary tables of missing information at the marker, individual and population level are generated. Below, the simplest form of the function to get results. More options are available, please see the function documentation.</p>
</div>
<div id="prepare-your-r-workspace" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-your-r-workspace" class="anchor"></a>Prepare your R workspace</h1>
<p>Clean your desk</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(<span class="dt">list =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>())</a></code></pre></div>
<p>Follow the <a href="https://thierrygosselin.github.io/grur/">instruction to install grur</a></p>
<p>Load the required libraries:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"grur"</span>)</a></code></pre></div>
<p>Set your working directory, e.g.:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"~/Documents/test_missing_visualization_vignette"</span>)</a></code></pre></div>
<p>Note: running codes in chunks inside R Notebook might cause problem, run it outside in the console (the default here).</p>
</div>
<div id="download-the-test-data" class="section level1">
<h1 class="hasAnchor">
<a href="#download-the-test-data" class="anchor"></a>Download the test data</h1>
<p>Dataset: in this example, we use the data in Ferchaud and Hansen (2015 and 2016) paper. The code below gets the vcf from Dryad directly. But you can skip the step if it’s already in the folder.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readBin">writeBin</a></span>(httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/content">content</a></span>(httr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/httr/topics/GET">GET</a></span>(<span class="st">"http://datadryad.org/bitstream/handle/10255/dryad.97237/sticklebacks_Danish.vcf?sequence=1"</span>), <span class="st">"raw"</span>), <span class="st">"stickleback_ferchaud_2015.vcf"</span>)</a></code></pre></div>
<p>With a vcf you also need a strata file (indicating population groupings)</p>
<ul>
<li>it’s a tab-delimited file or object, in the global environment.</li>
<li>requires a minimum of 2 columns: <code>INDIVIDUALS</code> and <code>STRATA</code>.</li>
<li>the <code>STRATA</code> column and the remaining columns can be any hierarchical groupings you like.</li>
<li>because in the strickleback vcf the population’s ids are contained in the name of the individuals, we can easily extract this info with string command.</li>
<li>here, creating from scratch the strata file is beyond the point of this tutoria (<code>??radiator::read_strata</code>).</li>
<li>if you don’t have this kind of individual naming scheme, you can also make the strata file by hand, the old fashion way.</li>
<li>if you’ve used  for your RADseq pipeline, the strata file is similar to a stacks <code>population map file</code>. Just make sure you have the required column names: <code>INDIVIDUALS</code> and <code>STRATA</code>.</li>
</ul>
<p>to download the strata for this example: <a href="https://www.dropbox.com/s/ely3wp4j4tulkrc/strata.stickleback.tsv?dl=0">strata link</a></p>
</div>
<div id="run-the-function" class="section level1">
<h1 class="hasAnchor">
<a href="#run-the-function" class="anchor"></a>Run the function</h1>
<p>This is the simplest way to run the function:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">ibm &lt;-<span class="st"> </span>grur<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/grur/topics/missing_visualization">missing_visualization</a></span>(</a>
<a class="sourceLine" id="cb5-2" title="2">  <span class="dt">data =</span> <span class="st">"stickleback_ferchaud_2015.vcf"</span>, </a>
<a class="sourceLine" id="cb5-3" title="3">  <span class="dt">strata =</span> <span class="st">"strata.stickleback.tsv"</span>)</a></code></pre></div>
<p>The function does a few automatic filters: * Monomorphic markers are removed * Only common markers between strata are kept for the analysis * Individuals and markers statistics are generated automatically</p>
<p>A new object <code>ibm</code> was created in your global environment. It’s a list and to view it’s content use:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(ibm)</a></code></pre></div>
<p>Lots of info in there… Lets focus on just a few. A folder is also created automatically. The function generates by default a large object (list):</p>
<ul>
<li>IBM-PCoA plot</li>
<li>heatmap plot showing genotyped/missing data (in black)</li>
<li>several plots showing the distribution of missing genotypes by individuals, pop and markers.</li>
<li>blacklists of individuals are included in the list <code>ibm</code> and written to the working directory (defaults from 10% to 70%, that automatically stops if no individuals at the threshold is found, can be changed with <code>ind.missing.geno.threshold</code> argument).</li>
<li>tables with summary missing information along the vectors of eigenvalues of the principal coordinates analysis are also part of the list output.</li>
<li>all these objects are accessed with <code>$</code>.</li>
</ul>
</div>
<div id="visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#visualization" class="anchor"></a>Visualization</h1>
<p><strong>To view the IBM-PCoA plot made with POP_ID grouping:</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">ibm<span class="op">$</span>ibm.plots<span class="op">$</span>ibm.strata.POP_ID</a></code></pre></div>
<p>The dark green bubble from KIB it’s an individual with almost <strong>all of his genotypes missing</strong>. This one skip the radar of the authors ;)</p>
<p>The heatmap showing missingness:</p>
<ul>
<li>gives a nice global outlook of the missingness</li>
<li>the figure is very long to generate (on some computers..), don’t try this with more than 100K markers… or take a couple day off:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">heatmap &lt;-<span class="st"> </span>ibm<span class="op">$</span>heatmap </a>
<a class="sourceLine" id="cb8-2" title="2">heatmap </a></code></pre></div>
<p>The vertical black line highlight the problem in the vcf with the individual missing almost all it’s genotypes.</p>
<p>View the table with summary of missing genotypes per individuals:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">table.ind &lt;-<span class="st"> </span>ibm<span class="op">$</span>missing.genotypes.ind</a>
<a class="sourceLine" id="cb9-2" title="2">table.ind</a></code></pre></div>
<p>To view the distribution</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">ibm<span class="op">$</span>missing.genotypes.ind.plots</a></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">ibm<span class="op">$</span>missing.genotypes.ind.histo</a></code></pre></div>
<p>Show the helper figure showing how many individuals could potentially be blacklisted based on % on genotypes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">ibm<span class="op">$</span>ind.genotyped.helper.plot</a></code></pre></div>
<p>All these figures are combined in the folder…</p>
<p>To view the distribution of missingness per markers</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">ibm<span class="op">$</span>missing.genotypes.markers.combined.plots</a></code></pre></div>
<p><strong>Other figures are created, explore the list of objects and folder. Read the doc.</strong></p>
<p><strong>To view the distribution of FH and missing genotypes per individuals</strong></p>
<ul>
<li><p>FH : is an identity by descent genomic (IBDg) measure based on the excess in the observed number of homozygous genotypes within an individual relative to the mean number of homozygous genotypes expected under random mating).</p></li>
<li><p>To learn more about it see (Keller et al., 2011; Kardos et al., 2015; Hedrick &amp; Garcia-Dorado, 2016)</p></li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">ibm<span class="op">$</span>missing.genotypes.ind.fh.combined.plots </a></code></pre></div>
<p>This is weird figure is caused by the outlier individual. To remove this individual, re-run <code>missing_visualization</code> with the argument <code>blacklist.id</code> and one of the several blacklists written to the working directory (e.g. <code>blacklist.id.missing.70.tsv</code>).</p>
<p><strong>Explore the rest by yourself!</strong></p>
</div>
<div id="interpretation" class="section level1">
<h1 class="hasAnchor">
<a href="#interpretation" class="anchor"></a>Interpretation</h1>
<p>Do you see patterns in your plots that provides insight about the relationships that missing values might have with other variables (inspired from <a href="http://r4ds.had.co.nz/exploratory-data-analysis-eda.html">r4ds</a>).</p>
<p>If you see a pattern, ask yourself:</p>
<ul>
<li><p>Is the pattern due to coincidence (i.e. random chance)?</p></li>
<li><p>Could you describe the relationship in the pattern ?</p></li>
<li><p>How strong is the relationship implied by the pattern?</p></li>
<li><p>What other variables might affect the relationship?</p></li>
<li><p>Does the relationship change if you look at individual subgroups of the data?</p></li>
<li><p>Do you think the pattern observed in the data could impact the clustering analysis?</p></li>
</ul>
</div>
<div id="strategies" class="section level1">
<h1 class="hasAnchor">
<a href="#strategies" class="anchor"></a>Strategies</h1>
<div id="arguments" class="section level2">
<h2 class="hasAnchor">
<a href="#arguments" class="anchor"></a>Arguments</h2>
<ul>
<li>use <code>strata.select</code> to select columns from the strata file to generate PCoA-IBM plots. If you have several columns to test e.g. library, sequencer, sequencing lanes, sampling sites, populations, use inside a string to get all of them!</li>
<li>use <code>distance.method</code> to explore other distance metric used by the function <code>dist</code>
</li>
</ul>
</div>
<div id="filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h2>
<ul>
<li>explore the impact of different filtering parameters on missing genotypes pattern, with <code><a href="https://www.rdocumentation.org/packages/radiator/topics/filter_rad">radiator::filter_rad</a></code>.</li>
<li>test your own whitelist of markers with <code>whitelist.markers</code>.</li>
<li>use the blacklist of individuals created with the function to <strong>manage</strong> missing data inside filtering pipelines</li>
<li>excluding/including individuals, populations and markers to test the impact of filtering on polymorphism discovery</li>
<li>ready to test missing data imputations ? Most stackr functions have built-in imputations arguments.</li>
</ul>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>Ferchaud A, Hansen MM (2016) The impact of selection, gene flow and demographic history on heterogeneous genomic divergence: threespine sticklebacks in divergent environments. Molecular Ecology 25(1): 238–259. <a href="http://dx.doi.org/10.1111/mec.13399" class="uri">http://dx.doi.org/10.1111/mec.13399</a></p>
<p>Ferchaud A, Hansen MM (2015) Data from: The impact of selection, gene flow and demographic history on heterogeneous genomic divergence: threespine sticklebacks in divergent environments. Dryad Digital Repository. <a href="http://dx.doi.org/10.5061/dryad.kp11q" class="uri">http://dx.doi.org/10.5061/dryad.kp11q</a></p>
<p>Danecek P, Auton A, Abecasis G et al. (2011) The variant call format and VCFtools. Bioinformatics, 27, 2156–2158.</p>
<p>Purcell S, Neale B, Todd-Brown K et al. (2007) PLINK: a tool set for whole-genome association and population-based linkage analyses. American Journal of Human Genetics, 81, 559–575.</p>
<p>Keller MC, Visscher PM, Goddard ME (2011) Quantification of inbreeding due to distant ancestors and its detection using dense single nucleotide polymorphism data. Genetics, 189, 237–249.</p>
<p>Kardos M, Luikart G, Allendorf FW (2015) Measuring individual inbreeding in the age of genomics: marker-based measures are better than pedigrees. Heredity, 115, 63–72.</p>
<p>Hedrick PW, Garcia-Dorado A. (2016) Understanding Inbreeding Depression, Purging, and Genetic Rescue. Trends in Ecology and Evolution. 2016;31: 940-952. <a href="doi:10.1016/j.tree.2016.09.005" class="uri">doi:10.1016/j.tree.2016.09.005</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#objectives">Objectives</a></li>
      <li><a href="#workflow">Workflow</a></li>
      <li><a href="#prepare-your-r-workspace">Prepare your R workspace</a></li>
      <li><a href="#download-the-test-data">Download the test data</a></li>
      <li><a href="#run-the-function">Run the function</a></li>
      <li><a href="#visualization">Visualization</a></li>
      <li><a href="#interpretation">Interpretation</a></li>
      <li>
<a href="#strategies">Strategies</a><ul class="nav nav-pills nav-stacked">
<li><a href="#arguments">Arguments</a></li>
      <li><a href="#filtering">Filtering</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Thierry Gosselin, Eric Archer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
