<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Memorize missingness pattern and randomize attributes — memorize_missing • grur</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Memorize missingness pattern and randomize attributes — memorize_missing" />
<meta property="og:description" content="Use this function to keep the pattern of missing data (0/1).
The pattern can be randomized based on dataset attributes/covariates.
This can be useful to generate missingness on simulated dataset with the
same number of individuals, populations and markers or to analyze the accuracy
of imputation algorithms. A vignette is under construction to leverage this
function." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">grur</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Installation</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#assumptions-before-imputing-your-dataset">Assumptions</a>
</li>
<li>
  <a href="https://thierrygosselin.github.io/grur/#features">Features</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rad_genomics_computer_setup.html">Computer setup - Installation - Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/vignette_missing_data_analysis.html">Missing visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/thierrygosselin/grur">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Memorize missingness pattern and randomize attributes</h1>
    <small class="dont-index">Source: <a href='https://github.com/thierrygosselin/grur/blob/master/R/memorize_missing.R'><code>R/memorize_missing.R</code></a></small>
    <div class="hidden name"><code>memorize_missing.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use this function to keep the pattern of missing data (0/1).
The pattern can be randomized based on dataset attributes/covariates.
This can be useful to generate missingness on simulated dataset with the
same number of individuals, populations and markers or to analyze the accuracy
of imputation algorithms. A vignette is under construction to leverage this
function.</p>
    </div>

    <pre class="usage"><span class='fu'>memorize_missing</span><span class='op'>(</span><span class='va'>data</span>, strata <span class='op'>=</span> <span class='cn'>NULL</span>, randomize <span class='op'>=</span> <span class='cn'>NULL</span>, filename <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>14 options for input (<strong>diploid data only</strong>): VCFs (SNPs or Haplotypes,
to make the vcf population ready),
plink (tped, bed), stacks haplotype file, genind (library(adegenet)),
genlight (library(adegenet)), gtypes (library(strataG)), genepop, DArT,
and a data frame in long/tidy or wide format. To verify that radiator detect
your file format use <code><a href='https://thierrygosselin.github.io/radiator/reference/detect_genomic_format.html'>detect_genomic_format</a></code> (see example below).
Documented in <strong>Input genomic datasets</strong> of <code><a href='https://thierrygosselin.github.io/radiator/reference/tidy_genomic_data.html'>tidy_genomic_data</a></code>.</p>
<p><strong>DArT and VCF data</strong>: <span class="pkg">radiator</span> was not meant to generate alleles
and genotypes if you are using a VCF file with no genotype
(only genotype likelihood: GL or PL).
Neither is <span class="pkg">radiator</span> able to magically generate a genind object
from a SilicoDArT dataset. Please look at the first few lines of your dataset
to understand it's limit before asking raditor to convert or filter your dataset.</p></td>
    </tr>
    <tr>
      <th>strata</th>
      <td><p>(optional/required) Required for VCF and haplotypes files,
optional for the other formats supported.
See documentation of <code><a href='https://thierrygosselin.github.io/radiator/reference/tidy_genomic_data.html'>tidy_genomic_data</a></code> for more info.
Default: <code>strata = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>randomize</th>
      <td><p>(optional, string) To randomize the missingness of specific attributes.
Available options: <code>"markers", "populations", "individuals" and "overall"</code>.
All options can be selected in a string,
<code>randomize = c("markers", "populations", "individuals", "overall")</code>
Default: <code>randomize = NULL</code> will only keep the original missingness pattern.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>(optional) The name of the file (extension not necessary)
written to the working directory and containing the missing info.
Default: <code>filename = NULL</code>, the missing info is in the global
environment only.</p>
<p><code>grur</code> takes advantage of the lightweight and speedy file reading/writing
package <code>fst</code> (<em>Lightning Fast Serialization of Data Frames for R</em>)
to write the dataframe to the working directory. This file can be used inside
<code>generate_missing</code> (coming soon) function.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A tidy dataframe in the global environment with columns: 
<code>POP_ID</code>, <code>INDIVIDUALS</code>, <code>MARKERS</code>, and in the subsequent
columns, the missingness info coded 0 for missing and 1 for genotyped.
Depending on the value chosen for the argument <code>randomize</code>, 
the columns are:</p><ul>
<li><p><code>MISSING_ORIGINAL</code>: for the original missing pattern (always present)</p></li>
<li><p><code>MISSING_MARKERS_MIX</code>: for the missing pattern randomized by markers (optional)</p></li>
<li><p><code>MISSING_POP_MIX</code>: for the missing pattern randomized by populations (optional)</p></li>
<li><p><code>MISSING_INDIVIDUALS_MIX</code>: for the missing pattern randomized by individuals (optional)</p></li>
<li><p><code>MISSING_OVERALL_MIX</code>: for the missing pattern randomized overall (optional)</p></li>
</ul>

    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Thierry Gosselin <a href='mailto:thierrygosselin@icloud.com'>thierrygosselin@icloud.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>missing.memory</span> <span class='op'>&lt;-</span> <span class='fu'>memorize_missing</span><span class='op'>(</span>
data <span class='op'>=</span> <span class='st'>"batch_1.vcf"</span>,
strata <span class='op'>=</span> <span class='st'>"population.map.strata.tsv"</span>, 
randomize <span class='op'>=</span> <span class='st'>"populations"</span>, filename <span class='op'>=</span> <span class='st'>"missing.memory.panda"</span>
<span class='op'>)</span>
<span class='op'>}</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Thierry Gosselin, Eric Archer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


